receiver/snmpreceiver/client.go

```diff
@@ -0,0 +1,539 @@
+type oidDataType byte
+
+const (
+	NotSupported oidDataType = 0x00
+	Integer      oidDataType = 0x01 // value will be int64
+	Float        oidDataType = 0x02 // value will be float64
+	String       oidDataType = 0x03 // value will be string
+)
+
+// snmpData used for processFunc and is a simpler version of gosnmp.SnmpPDU
+type snmpData struct {
+	parentOID string // optional
+	oid       string
+	value     interface{}
+	valueType oidDataType
+}
+
+// processFunc our own function type to better control what data can be passed in
+type processFunc func(data snmpData) error
+
+// client is used for retrieving data from a SNMP environment
+type client interface {
+	// GetScalarData retrieves SNMP scalar data from a list of passed in OIDS,
+	// then the passed in function is performed on each piece of data
+	GetScalarData(oids []string, processFn processFunc) error
+	// GetIndexedData retrieves SNMP indexed data from a list of passed in OIDS,
+	// then the passed in function is performed on each piece of data
+	GetIndexedData(oids []string, processFn processFunc) error
+	// Connect makes a connection to the SNMP host
+	Connect() error
+	// Close closes a connection to the SNMP host
+	Close() error
+}
+
+// snmpClient implements the client interface and retrieves data through SNMP
+type snmpClient struct {
+	client goSNMPWrapper
+	logger *zap.Logger
+}
+
+// Verify snmpClient implements client interface
+var _ client = (*snmpClient)(nil)
+
+// newClient creates an initialized client
+func newClient(cfg *Config, host component.Host, settings component.TelemetrySettings, logger *zap.Logger) (client, error) {
+	// Create goSNMP client
+	goSNMP := newGoSNMPWrapper()
+	goSNMP.SetTimeout(cfg.CollectionInterval)
+
+	// Set goSNMP version based on config
+	switch cfg.Version {
+	case "v3":
+		goSNMP.SetVersion(gosnmp.Version3)
+	case "v2c":
+		goSNMP.SetVersion(gosnmp.Version2c)
+	case "v1":
+		goSNMP.SetVersion(gosnmp.Version1)
+	default:
+		return nil, errors.New("failed to create goSNMP client: invalid version")
+	}
+
+	// Verify config endpoint is good
+	endpoint := cfg.Endpoint
+	if !strings.Contains(endpoint, "://") {
+		endpoint = "udp://" + endpoint
+	}
+
+	snmpUrl, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create goSNMP client: invalid endpoint. %w", err)
+	}
+
+	// Set goSNMP transport based on config
+	switch snmpUrl.Scheme {
+	case "tcp", "tcp4", "tcp6", "udp", "udp4", "udp6":
+		goSNMP.SetTransport(snmpUrl.Scheme)
+	default:
+		return nil, fmt.Errorf("failed to create goSNMP client: unsupported scheme '%s'", snmpUrl.Scheme)
+	}
+
+	// Set goSNMP port based on config
+	portStr := snmpUrl.Port()
+	if portStr == "" {
+		portStr = "161"
+	}
+
+	port, err := strconv.ParseUint(portStr, 10, 16)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create goSNMP client: issue parsing port '%s'. %w", portStr, err)
+	}
+	goSNMP.SetPort(uint16(port))
```

Comment:
There's a lot of validation in here that seems like it should be handled at the config validation level.

Reply:
@cpheps i remember thinking the same thing. i think a fair amount of it is "double checking", but some might be unique. let me take another look at this.

Reply:
@cpheps see above, removed any client validation that is done in config